### PR TITLE
Currency configuration

### PR DESCRIPTION
**Pull request by veylib**

### CODE REVIEW

File: src/Calculator/StandardCalculator.php [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        if ($left->isEmpty()) {
+            return clone $right;
+        }
+        if ($right->isEmpty()) {
+            return clone $left;
+        }
+        $this->compareCurrencySymbols($left->getCurrency(), $right->getCurrency());
+        $currency = $this->buildCurrency($left, $right);
+        $newGross = $left->getGross() + $right->getGross();
+        $newNett = $left->getNett() + $right->getNett();
```
Comment (maureis) [inline src/Calculator/StandardCalculator.php:34]: strategy should be used here to calculate new nett from gross or gross from nett. Gross or nett (depends on project strategy) prices has round error so operations add/subtract/multiply/divide are incresing that error.

File: tests/Calculator/StandardCalculatorTest.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+     * @dataProvider subtractDataProvider
+     */
+    public function testSubtract(float $lValue, string $lCurrency, float $rValue, string $rCurrency, float $eResult)
+    {
+        $calculator = new StandardCalculator();
+        $result = $calculator->subtract(
+            Price::buildByNett($lValue, 0, $lCurrency),
+            Price::buildByNett($rValue, 0, $rCurrency)
+        );
+        $this->assertEquals($eResult, $result->getNett());
```
Comment (maureis) [inline tests/Calculator/StandardCalculatorTest.php]: gross also should be tested (all operations) for the round error case

File: src/Calculator/StandardCalculator.php (hunk 1 above)
Comment (veylib) [reply src/Calculator/StandardCalculator.php:34]: I would not change current logic - I just moved it outside from price https://share.merce.com/files/jgxmx4dip.png

File: tests/Calculator/StandardCalculatorTest.php (hunk 2 above)
Comment (veylib) [reply tests/Calculator/StandardCalculatorTest.php]: Added a lot of test cases and added check for gross/nett/tax values.

File: .phpunit.result.cache [hunk 3]
```diff
@@ -0,0 +1 @@
...et #9";d:0;s:81:"Ayeo\Price\Test\Calculator\StandardCalculatorTest::testSubtract with data set #10";d:0;s:75:"Ayeo\Price\Test\Calculator\StandardCalculatorTest::testAdd with data set #4";d:0;s:80:"Ayeo\Price\Test\Calculator\StandardCalculatorTest::testMultiply with data set #6";d:0;s:80:"Ayeo\Price\Test\Calculator\StandardCalculatorTest::testMultiply with data set #7";d:0;s:80:"Ayeo\Price\Test\Calculator\StandardCalculatorTest::testMultiply with data set #8";d:0;s:80:"Ayeo\Price\Test\Calculator\StandardCalculatorTest::testMultiply with data set #9";d:0;s:81:"Ayeo\Price\Test\Calculator\StandardCalculatorTest::testMultiply with data set #10";d:0;s:81:"Ayeo\Price\Test\Calculator\StandardCalculatorTest::testMultiply with data set #11";d:0;s:78:"Ayeo\Price\Test\Calculator\StandardCalculatorTest::testDivide with data set #7";d:0;s:78:"Ayeo\Price\Test\Calculator\StandardCalculatorTest::testDivide with data set #8";d:0;s:78:"Ayeo\Price\Test\Calculator\StandardCalculatorTest::testDivide with data set #9";d:0;s:79:"Ayeo\Price\Test\Calculator\StandardCalculatorTest::testDivide with data set #10";d:0;s:79:"Ayeo\Price\Test\Calculator\StandardCalculatorTest::testDivide with data set #11";d:0;}}}
```
Comment (ayeo) [inline .phpunit.result.cache]: this file should be ignored imo

File: src/Calculator/StandardCalculator.php [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    public function __construct(DecoratorInterface ...$decorators)
+    {
+        $this->decorators = $decorators;
+    }
+
+    public function add(Price $left, Price $right): Price
+    {
+        if ($left->isEmpty()) {
+            return clone $right;
+        }
```
Comment (ayeo) [inline src/Calculator/StandardCalculator.php:25]: blank line after condition

File: src/Price.php [hunk 5]
```diff
@@ -1,55 +1,31 @@
...
-
+        }
+        elseif($currencySymbol !== null)
+        {
+            $this->currency = new Currency($currencySymbol);
+        }
 
-        $this->nett = new Money($nett);
-        $this->gross = new Money($gross);
+        $this->nett = $this->getCalculator($this->currency)->decorateMoney(new Money($nett));
```
Comment (ayeo) [inline src/Price.php]: not sure if extracting Money make much of sense. It looks to me like in ALL cases you will ALWAYS decorate nett/gross same way - so decorator shoult work with Price itself imo (just return new Price). Actualy it is not the decorator it is more Chain of Responsibility ;)

File: src/Calculator/StandardCalculator.php (hunk 4 above)
Comment (veylib) [reply src/Calculator/StandardCalculator.php:25]: Fixed.

File: .phpunit.result.cache (hunk 3 above)
Comment (veylib) [reply .phpunit.result.cache]: Yeah, fixed.

File: src/Price.php (hunk 5 above)
Comment (veylib) [reply src/Price.php]: Changed to decorating price.

It's not a decorate pattern itself, it just apply some effects on price.

Comment (ayeo) [reply src/Price.php]: so why it is called decorator? :D

File: src/Money.php [hunk 6]
```diff
@@ -1,42 +1,26 @@
 <?php
+
 namespace Ayeo\Price;
 
 class Money
```
Comment (ayeo) [inline src/Money.php:5]: To me such money concept is worthless. It does not reflect any real knowledge. I think money consist of float value and currency. The 100.0 is not money at all. I can not see any advantages of introducing such idea here. The difference between price and money is only lack of tax probably but such distinction has no value itself in the model. The one scenario I can imagine is zero-price which some may argue is not price at all - but at the second thought it is no money either :D Maybe is is worth to introduce ZeroPrice singleton - excelent example of singleton use case scenario - but this is topic for another story

**Pull request merged**